# 🧹 Day 5 — Data Cleaning

**Roadmap:** Finding blanks → Deleting → Filling → Extreme values → ☕ → Transforming → Duplicates → Saving → 🚢 Titanic

📂 **Datasets:** `planets.csv` and `Titanic-Dataset.csv` — keep both in this folder.

In [2]:
import numpy as np 
import pandas as pd

In [4]:

p=pd.read_csv("/Users/ankit/BOX 1/B1 ROOM 1/AI&ML/class7/planets.csv")

# 🔁 Revision — Day 4

Same file as last class. Nothing new here — every question uses something
you already wrote in last class.



**1.** The archive wants a list of the oldest people on board.
Show only `Name`, `Pclass` and `Fare` for passengers older than 65.
How many are there?

**2.** How many passengers did **not** board at Southampton? Use `~`.




In [5]:
p.head(1)

,method,number,orbital_period,mass,distance,year
0,Radial Velocity,1,269.3,7.1,77.4,2006


247

**3.** How many passengers were either under 10 or over 70? Use `|`.

**4.** How many passengers travelled in each class?

Then, among **third class only**, how many were male and how many female?

**5.** Create a column called `travel_group`. Everyone starts as `'alone'`.
Anyone whose `SibSp` plus `Parch` is more than 0 becomes `'with family'`.

In [6]:
p.head(1)

,method,number,orbital_period,mass,distance,year
0,Radial Velocity,1,269.3,7.1,77.4,2006


### Yesterday

150 passengers over 40, 564 aged 40 or under. Total **714**, not 891.
177 rows were dropped from every answer and nothing warned us.

Three options exist, and each one changes the answer: **delete**, **fill**, or **leave it**.
Today we do all three.

In [ ]:
p.shape

**1035 planets discovered outside our solar system, 1989–2014.**

| Column | Meaning |
|---|---|
| `method` | how it was detected |
| `number` | planets known in that star system |
| `orbital_period` | days to orbit once |
| `mass` | in Jupiter masses |
| `distance` | from Earth, in parsecs |
| `year` | year of discovery |

---
## 1️⃣ Finding blanks

> **Flow:** one cell → one column → the whole table

In [7]:
# True wherever a value is missing

p.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1035 entries, 0 to 1034
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   method          1035 non-null   object 
 1   number          1035 non-null   int64  
 2   orbital_period  992 non-null    float64
 3   mass            513 non-null    float64
 4   distance        808 non-null    float64
 5   year            1035 non-null   int64  
dtypes: float64(3), int64(2), object(1)
memory usage: 48.6+ KB


In [9]:
# count the blanks in one column
p.isnull().sum()

method              0
number              0
orbital_period     43
mass              522
distance          227
year                0
dtype: int64

In [10]:
# same information, different view
p["mass"].isnull().sum()

np.int64(522)

| Column | Blanks |
|---|---|
| `mass` | 522 |
| `distance` | 227 |
| `orbital_period` | 43 |
| `method`, `number`, `year` | 0 |

`isnull()` and `isna()` are the same function — two names for one thing.
`notnull()` is its opposite.

### ✋ Guess first

`p.dropna()` removes every row with any blank in it. We have **1035** planets.





In [11]:
#
p.dropna()

,method,number,orbital_period,mass,distance,year
0,Radial Velocity,1,269.30000,7.100,77.40,2006
1,Radial Velocity,1,874.77400,2.210,56.95,2008
2,Radial Velocity,1,763.00000,2.600,19.84,2011
3,Radial Velocity,1,326.03000,19.400,110.62,2007
4,Radial Velocity,1,516.22000,10.500,119.47,2009
...,...,...,...,...,...,...
640,Radial Velocity,1,111.70000,2.100,14.90,2009
641,Radial Velocity,1,5.05050,1.068,44.46,2013
642,Radial Velocity,1,311.28800,1.940,17.24,1999
649,Transit,1,2.70339,1.470,178.00,2013


**498 of 1035.** Half the catalogue, and most of it went because of `mass` —
a column you may not even have needed.

In [ ]:
# drop only where mass is blank
p.dropna

TypeError: DataFrame.dropna() takes 1 positional argument but 2 were given

In [ ]:
# drop the damaged columns instead, and see which survive



| Command | Result |
|---|---|
| `p.dropna()` | 498 rows |
| `p.dropna(subset=['mass'])` | 513 rows |
| `p.dropna(subset=['distance'])` | 808 rows |
| `p.dropna(axis=1)` | 3 columns — only `method`, `number`, `year` |

⚠️ No `subset` means **any** blank anywhere kills the row.

### 🧩 Try these — on the **Titanic** file


**2.1** Two Titanic passengers have no boarding port recorded. Drop only those rows.
How many passengers remain?


In [16]:
# 2.1

df=pd.read_csv("/Users/ankit/BOX 1/B1 ROOM 1/AI&ML/Class6/test.csv")
df.head()


,PassengerId,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,892,3,"Kelly, Mr. James",male,34.5,0,0,330911,7.8292,NaN,Q
1,893,3,"Wilkes, Mrs. James (Ellen Needs)",female,47.0,1,0,363272,7.0000,NaN,S
2,894,2,"Myles, Mr. Thomas Francis",male,62.0,0,0,240276,9.6875,NaN,Q
3,895,3,"Wirz, Mr. Albert",male,27.0,0,0,315154,8.6625,NaN,S
4,896,3,"Hirvonen, Mrs. Alexander (Helga E Lindqvist)",female,22.0,1,1,3101298,12.2875,NaN,S


**2.2** Keep all 891 passengers, but drop every **column** that has a blank.
How many columns survive, and which three did you lose?

In [ ]:
# 2.2

df["Fare"]

---
## 3️⃣ Filling

```python
p['mass'].fillna(0)                    # a fixed value
p['mass'].fillna(p['mass'].mean())     # the average
p['mass'].fillna(p['mass'].median())   # the middle value
```

> **Flow:** mean vs median → fill with each → what the fill did to the data

In [17]:
# mean and median of mass
p['mass'].fillna(0).isnull().sum()

np.int64(0)

**Mean 2.64, median 1.26.** The mean is double the median — a few very heavy planets pull it up.

In [ ]:
# fill mass with the mean, then check mean / median / std

| | before | after mean-fill |
|---|---|---|
| mean | 2.64 | 2.64 |
| **median** | **1.26** | **2.64** |
| **std** | **3.82** | **2.69** |

522 identical values landed in the middle of the column. The median moved to sit on top of them
and the spread shrank.

### 🧩 Try these — on the **Titanic** file

The demo above was on `planets`. These are the same moves on `t`, so you are
practising the idea and not repeating the cells.

**3.1** Work on a copy first: `t2 = t.copy()`

Fill the missing `Age` in `t2` with the **mean**, then count how many passengers are now
recorded as exactly **29.6991** years old. Before you ran it, that count was **0**.

In [ ]:
# 3.1
# t2 = t.copy()   <- do not touch t, we need its blanks later

**3.2** `Fare` has mean **32.20** and median **14.45** — more than double apart.

If `Fare` had blanks, which of the two would you fill with? One line, no code.

In [ ]:
# 3.2

---
## 5️⃣ Duplicates

```python
p.duplicated()                        # True for a row that repeats an earlier one
p.duplicated().sum()                  # how many
p.drop_duplicates()                   # remove them
p.duplicated(subset=['method'])       # "same" only in the columns you name
```

> **Flow:** find → count → show → remove → then decide what "the same" means

In [ ]:
# how many exact duplicate rows?

In [ ]:
# show them

In [ ]:
# remove them and check the row count

**4 exact duplicates → 1031 rows.** Two Imaging pairs and two Microlensing pairs.

⚠️ `subset=` changes the whole meaning:

In [ ]:
# duplicates by method only

**1025.** There are only 10 detection methods for 1035 planets, so almost every row repeats one.
Those are not errors.

**A duplicate is only a duplicate relative to the columns you name.**

---
## 💾 Saving your work

Cleaning that is not saved is cleaning that did not happen.

```python
p.to_csv('planets_clean.csv', index=False)
```

> **Flow:** save → read it back → check nothing changed

In [ ]:
# save the cleaned table

In [ ]:
# read it back and look at the columns

⚠️ Leave out `index=False` and pandas writes the row labels as a real column.
Read that file back and you get an extra `Unnamed: 0` sitting in your table forever.

### 🧩 Try this — on the **Titanic** file

**B.1** Save the Titanic table as `titanic_clean.csv` with `index=False`,
read it back into a new variable, and confirm the columns are identical to `t`.

In [ ]:
# B.1

In [ ]:
t.head()

**T1** Which three columns are damaged, and by how much?

In [ ]:
# T1

**T2** Run `t.dropna()` with no arguments. How many of the 891 passengers survive?
Which column caused most of that damage?

In [ ]:
# T2

**T3** Fill `Embarked` with the most common port, and `Age` with the median.

In [ ]:
# T3

**T4** Build a `has_cabin` column and compare the survival rate of the two groups.

You did the same move on `Age` in 4.1. Same idea, different column — watch what changes.

In [ ]:
# T4

**66.7% against 30.0%.** The most damaged column in the file — 687 blanks out of 891 —
separates survivors better than almost anything else in it. A recorded cabin number meant
first class, upper deck, near the boats.

---
## 📌 Recap

### Finding
```python
p.isnull().sum()      # blanks per column   (isna() is the same function)
p.info()              # non-null counts and types
p.describe()          # min, median, max — where the extremes hide
```

### The three options

| Option | Code | Use it when |
|---|---|---|
| **Delete** | `p.dropna(subset=['mass'])` | few blanks, and you can spare the rows |
| **Fill** | `p['mass'].fillna(value)` | the blank is a genuine unknown |
| **Transform** | `p['mass'].notnull()` | the blank itself means something |

⚠️ `dropna()` with no `subset` deletes on **any** blank: 1035 → 498 here, 891 → 183 on Titanic.

### Filling honestly
- **median** for numbers with a long tail — the mean gets dragged by extremes
- **mode** for text — `t['Embarked'].mode()[0]`
- every fill narrows the spread and invents rows that never existed
- ⭐ ask **who** is missing before **what** to fill. On Titanic, 136 of the 177 missing ages
  were third class, so a single global fill hides that pattern instead of handling it.

### Duplicates and saving
```python
p.duplicated().sum()                 # exact copies only
p.drop_duplicates()
p.duplicated(subset=['method'])      # "same" is whatever you define it to be
p.to_csv('clean.csv', index=False)   # index=False, always
```

### 🔜 Day 6
We just wrote three separate filters to get one number per detection method.
Tomorrow, one line does all ten at once.

---
## 🧠 Exit quiz — answer in chat

1. `p.dropna()` left 498 of 1035. Which column caused most of it?
2. You fill a column with its mean. What happens to the median, and to the std?
3. When is filling a column the wrong move entirely?
4. `p.duplicated(subset=['method'])` says 1025. Delete them or not?
5. What does `index=False` prevent?

In [ ]:
# quiz answers